File: testsuite/data/scores/multisig_wallet/multisig_wallet.py
# -*- coding: utf-8 -*-

# Copyright 2018 ICON Foundation
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

from .type_converter.type_converter import params_type_converter
from .qualification_check.qualification_check import *
from .transaction import Transaction


class MultiSigWallet(IconScoreBase):
    _MAX_WALLET_OWNER_COUNT = 50
    _MAX_DATA_REQUEST_AMOUNT = 50

    @eventlog(indexed=2)
    def Confirmation(self, _sender: Address, _transactionId: int):
        pass

    @eventlog(indexed=2)
    def Revocation(self, _sender: Address, _transactionId: int):
        pass

    @eventlog(indexed=1)
    def Submission(self, _transactionId: int):
        pass

    @eventlog(indexed=1)
    def Execution(self, _transactionId: int):
        pass

    @eventlog(indexed=1)
    def ExecutionFailure(self, _transactionId: int):
        pass

    @eventlog(indexed=1)
    def Deposit(self, _sender: Address, _value: int):
        pass

    @eventlog(indexed=1)
    def DepositToken(self, _sender: Address, _value: int, _data: bytes):
        pass

    @eventlog(indexed=1)
    def WalletOwnerAddition(self, _walletOwner: Address):
        pass

    @eventlog(indexed=1)
    def WalletOwnerRemoval(self, _walletOwner: Address):
        pass

    @eventlog
    def RequirementChange(self, _required: int):
        pass

    def __init__(self, db: IconScoreDatabase) -> None:
        super().__init__(db)
        # store transaction instance as a serialized bytes
        # _transactions's key: transaction id(int type)
        self._transactions = DictDB("transactions", db, value_type=bytes)
        # store wallet owners' confirmations of each transaction
        # _confirmations's key: transaction id(int type), address(Address type)
        self._wallet_owners = ArrayDB("wallet_owners", db, value_type=Address)
        self._confirmations = DictDB("confirmations", db, value_type=bool, depth=2)
        self._required = VarDB("required", db, value_type=int)
        self._transaction_count = VarDB('transactionCount', db, value_type=int)

    def on_install(self, _walletOwners: str, _required: int) -> None:
        super().on_install()

        wallet_owner_list = _walletOwners.replace(" ", "").split(",")
        self._check_requirement(len(wallet_owner_list), _required)

        for wallet_owner in wallet_owner_list:
            wallet_owner_address = Address.from_string(wallet_owner)
            self._wallet_owners.put(wallet_owner_address)

        self._required.set(_required)
        self._transaction_count.set(0)

    def on_update(self) -> None:
        super().on_update()

    @staticmethod
    def _check_params_format_convertible(json_formatted_params: str):
        # when user input None as a _params' value,
        # this will be changed to "" when creating Transaction instance.
        # "" will be changed to {} when finally execute transaction. so doesn't check format
        if json_formatted_params != "" and json_formatted_params is not None:
            try:
                params = json_loads(json_formatted_params)
                for param in params:
                    params_type_converter(param["type"], param["value"])
            except ValueError as e:
                revert(f"json format error: {e}")
            except IconScoreException as e:
                revert(f"{e}")
            except:
                revert("can not convert 'params' json data, check the 'params' parameter")

    @staticmethod
    def _only_positive_number(*args):
        for number in args:
            if number < 0:
                revert("only positive number is accepted")

    def _wallet_owner_does_not_exist(self, wallet_owner: Address):
        if wallet_owner in self._wallet_owners:
            revert(f"{wallet_owner} already exists as an owner of the wallet")

    def _wallet_owner_exist(self, wallet_owner: Address):
        if wallet_owner not in self._wallet_owners:
            revert(f"{wallet_owner} is not an owner of wallet")

    def _transaction_exists(self, transaction_id: int):
        if self._transactions[transaction_id] is None \
                or self._transaction_count.get() <= transaction_id:
            revert(f"transaction id '{transaction_id}' is not exist")

    def _confirmed(self, transaction_id: int, wallet_owner: Address):
        if not self._confirmations[transaction_id][wallet_owner]:
            revert(f"{wallet_owner} has not confirmed to the transaction id '{transaction_id}' yet")

    def _not_confirmed(self, transaction_id: int, wallet_owner: Address):
        if self._confirmations[transaction_id][wallet_owner]:
            revert(f"{wallet_owner} has already confirmed to the transaction '{transaction_id}'")

    def _not_executed(self, transaction_id: int):
        # before call this method, check if transaction is exists(use transaction_exists method)
        if self._transactions[transaction_id][0] == 1:
            revert(f"transaction id '{transaction_id}' has already been executed")

    def _check_requirement(self, wallet_owner_count: int, required: int):
        if wallet_owner_count > self._MAX_WALLET_OWNER_COUNT or \
                required > wallet_owner_count or \
                required <= 0 or \
                wallet_owner_count == 0:
            revert("invalid requirement")

    @payable
    def fallback(self):
        if self.msg.value > 0:
            self.Deposit(self.msg.sender, self.msg.value)

    @external
    def tokenFallback(self, _from: Address, _value: int, _data: bytes):
        if _value > 0:
            self.DepositToken(_from, _value, _data)

    @external
    def submitTransaction(self, _destination: Address,
                          _method: str = "", _params: str = "", _value: int = 0, _description: str = ""):
        self._wallet_owner_exist(self.msg.sender)
        # prevent failure of executing transaction caused by 'params' conversion problems
        self._check_params_format_convertible(_params)
        self._only_positive_number(_value)

        # add transaction
        transaction_id = self._add_transaction(_destination, _method, _params, _value, _description)
        # confirm_transaction
        self.confirmTransaction(transaction_id)

    @external
    def confirmTransaction(self, _transactionId: int):
        self._wallet_owner_exist(self.msg.sender)
        self._transaction_exists(_transactionId)
        self._not_confirmed(_transactionId, self.msg.sender)

        self._confirmations[_transactionId][self.msg.sender] = True

        self.Confirmation(self.msg.sender, _transactionId)

        self._execute_transaction(_transactionId)

    @external
    def revokeTransaction(self, _transactionId: int):
        self._wallet_owner_exist(self.msg.sender)
        self._transaction_exists(_transactionId)
        self._not_executed(_transactionId)
        self._confirmed(_transactionId, self.msg.sender)

        self._confirmations[_transactionId][self.msg.sender] = False

        self.Revocation(self.msg.sender, _transactionId)

    def _add_transaction(self, destination: Address, method: str, params: str, value: int, description: str) -> int:
        transaction = Transaction.create_transaction_with_validation(destination=destination,
                                                                     method=method,
                                                                     params=params,
                                                                     value=value,
                                                                     description=description)
        transaction_id = self._transaction_count.get()

        self._transactions[transaction_id] = transaction.to_bytes()
        self._transaction_count.set(transaction_id + 1)

        self.Submission(transaction_id)
        return transaction_id

    def _execute_transaction(self, transaction_id: int):
        # as this method can't be called from other SCORE or EOA, doesn't check owner, transactions_id, confirmations.
        if self._is_confirmed(transaction_id):
            if self._external_call(self._transactions[transaction_id]):
                self._transactions[transaction_id] = True.to_bytes(1, "big") + self._transactions[transaction_id][1:]

                self.Execution(transaction_id)
            else:
                self.ExecutionFailure(transaction_id)

    def _external_call(self, serialized_tx: bytes) -> bool:
        transaction = Transaction.from_bytes(serialized_tx)

        # if method == "" -> None
        method_name = None if transaction.method == "" else transaction.method
        # if params == "" -> {}
        method_params = {}
        if transaction.params != "":
            params = json_loads(transaction.params)
            for param in params:
                method_params[param["name"]] = params_type_converter(param["type"], param["value"])
        try:
            if transaction.destination.is_contract:
                self.call(addr_to=transaction.destination,
                          func_name=method_name,
                          kw_dict=method_params,
                          amount=transaction.value)
            else:
                self.icx.transfer(transaction.destination, transaction.value)
            execute_result = True
        except:
            execute_result = False

        return execute_result

    def _is_confirmed(self, transaction_id) -> bool:
        count = 0
        for wallet_owner in self._wallet_owners:
            if self._confirmations[transaction_id][wallet_owner]:
                count += 1

        return count == self._required.get()

    @only_wallet
    @external
    def addWalletOwner(self, _walletOwner: Address):
        self._wallet_owner_does_not_exist(_walletOwner)
        # check if owner's count exceed '_MAX_OWNER_COUNT'
        self._check_requirement(len(self._wallet_owners) + 1, self._required.get())

        self._wallet_owners.put(_walletOwner)

        self.WalletOwnerAddition(_walletOwner)

    @only_wallet
    @external
    def replaceWalletOwner(self, _walletOwner: Address, _newWalletOwner: Address):
        self._wallet_owner_exist(_walletOwner)
        self._wallet_owner_does_not_exist(_newWalletOwner)

        for idx, wallet_owner in enumerate(self._wallet_owners):
            if wallet_owner == _walletOwner:
                self._wallet_owners[idx] = _newWalletOwner
                break

        self.WalletOwnerRemoval(_walletOwner)
        self.WalletOwnerAddition(_newWalletOwner)

    @only_wallet
    @external
    def removeWalletOwner(self, _walletOwner: Address):
        self._wallet_owner_exist(_walletOwner)
        # if all owners are removed, this contract can not be executed.
        # so check if _owner is only one left in this wallet
        wallet_owners_count = len(self._wallet_owners)
        self._check_requirement(wallet_owners_count - 1, self._required.get())

        for idx, owner in enumerate(self._wallet_owners):
            if owner == _walletOwner:
                if idx == wallet_owners_count - 1:
                    self._wallet_owners.pop()
                else:
                    self._wallet_owners[idx] = self._wallet_owners.pop()
                break

        self.WalletOwnerRemoval(_walletOwner)

    @only_wallet
    @external
    def changeRequirement(self, _required: int):
        self._check_requirement(len(self._wallet_owners), _required)

        self._required.set(_required)

        self.RequirementChange(_required)

    @external(readonly=True)
    def getRequirement(self) -> int:
        return self._required.get()

    @external(readonly=True)
    def getTransactionInfo(self, _transactionId: int) -> dict:
        if self._transactions[_transactionId] is not None:
            transaction = Transaction.from_bytes(self._transactions[_transactionId])
            tx_dict = transaction.to_dict()
            tx_dict["_transactionId"] = _transactionId
            return tx_dict
        else:
            return {}

    @external(readonly=True)
    def getTransactionsExecuted(self, _transactionId: int) -> bool:
        if self._transactions[_transactionId] is not None:
            return bool(self._transactions[_transactionId][0])
        else:
            return False

    @external(readonly=True)
    def checkIfWalletOwner(self, _walletOwner: Address) -> bool:
        return _walletOwner in self._wallet_owners

    @external(readonly=True)
    def getWalletOwnerCount(self) -> int:
        return len(self._wallet_owners)

    @external(readonly=True)
    def getWalletOwners(self, _offset: int, _count: int) -> list:
        self._only_positive_number(_offset, _count)

        wallet_owner_list = []
        wallet_owners_count = len(self._wallet_owners)

        for idx in range(_offset, _offset + _count):
            if idx >= wallet_owners_count:
                break
            wallet_owner_list.append(str(self._wallet_owners[idx]))

        return wallet_owner_list

    @external(readonly=True)
    def getConfirmationCount(self, _transactionId: int) -> int:
        count = 0
        for wallet_owner in self._wallet_owners:
            if self._confirmations[_transactionId][wallet_owner]:
                count += 1
        return count

    @external(readonly=True)
    def getConfirmations(self, _offset: int, _count: int, _transactionId: int) -> list:
        self._only_positive_number(_offset, _count)

        confirmed_wallet_owners = []
        wallet_owners_count = len(self._wallet_owners)

        for idx in range(_offset, _offset + _count):
            if idx >= wallet_owners_count:
                break
            if self._confirmations[_transactionId][self._wallet_owners[idx]]:
                confirmed_wallet_owners.append(str(self._wallet_owners[idx]))

        return confirmed_wallet_owners

    @external(readonly=True)
    def getTransactionCount(self, _pending: bool = True, _executed: bool = True) -> int:
        tx_count = 0
        for tx_id in range(self._transaction_count.get()):
            if (_pending and not self._transactions[tx_id][0]) or (_executed and self._transactions[tx_id][0]):
                tx_count += 1

        return tx_count

    @external(readonly=True)
    def getTransactionList(self, _offset: int, _count: int, _pending: bool = True, _executed: bool = True) -> list:
        self._only_positive_number(_offset, _count)

        if _count > self._MAX_DATA_REQUEST_AMOUNT:
            revert("requests that exceed the allowed amount")

        transaction_list = []
        total_transaction_count = self._transaction_count.get()

        # prevent searching not existed transaction
        _count = _offset + _count if total_transaction_count >= _offset + _count else total_transaction_count

        for tx_id in range(_offset, _count):
            if (_pending and not self._transactions[tx_id][0]) or (_executed and self._transactions[tx_id][0]):
                transaction = Transaction.from_bytes(self._transactions[tx_id])

                tx_dict = transaction.to_dict()
                tx_dict["_transactionId"] = tx_id
                transaction_list.append(tx_dict)

        return transaction_list

File: javaee/samples/multisig/src/example/MultiSigWallet.java
/*
 * Copyright 2020 ICON Foundation
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package example;

import score.Address;
import score.ArrayDB;
import score.BranchDB;
import score.Context;
import score.DictDB;
import score.VarDB;
import score.annotation.EventLog;
import score.annotation.External;
import score.annotation.Optional;
import score.annotation.Payable;

import java.math.BigInteger;
import java.util.List;
import java.util.Map;

public class MultiSigWallet
{
    private static final int MAX_OWNER_COUNT = 50;

    private final ArrayDB<Address> owners = Context.newArrayDB("owners", Address.class);
    private final VarDB<BigInteger> required = Context.newVarDB("required", BigInteger.class);
    private final DictDB<BigInteger, Transaction> transactions = Context.newDictDB("transactions", Transaction.class);
    private final VarDB<BigInteger> transactionCount = Context.newVarDB("transactionCount", BigInteger.class);
    private final BranchDB<BigInteger, DictDB<Address, Boolean>>
            confirmations = Context.newBranchDB("confirmations", Boolean.class);

    /**
     * Contract constructor sets initial owners and required number of confirmations.
     */
    public MultiSigWallet(String _walletOwners, BigInteger _required) {
        assert(this.owners != null);
        StringTokenizer st = new StringTokenizer(_walletOwners, ", ");
        while (st.hasMoreTokens()) {
            this.owners.add(Address.fromString(st.nextToken()));
        }
        checkRequirement(this.owners.size(), _required);
        this.required.set(_required);
    }

    @Payable
    public void fallback() {
        BigInteger value = Context.getValue();
        if (value.signum() > 0) {
            Deposit(Context.getCaller(), value);
        }
    }

    @External
    public void tokenFallback(Address _from, BigInteger _value, byte[] _data) {
        if (_value.signum() > 0) {
            DepositToken(_from, _value, _data);
        }
    }

    @External
    public void addWalletOwner(Address _walletOwner) {
        onlyFromWallet();
        checkOwnerDoesNotExist(_walletOwner);
        checkRequirement(this.owners.size() + 1, this.required.get());
        // now we can add the owner
        this.owners.add(_walletOwner);
        WalletOwnerAddition(_walletOwner);
    }

    @External
    public void removeWalletOwner(Address _walletOwner) {
        onlyFromWallet();
        checkOwnerExist(_walletOwner);
        checkRequirement(this.owners.size() - 1, this.required.get());
        // get the topmost value
        Address top = this.owners.pop();
        if (!top.equals(_walletOwner)) {
            for (int i = 0; i < this.owners.size(); i++) {
                if (_walletOwner.equals(this.owners.get(i))) {
                    this.owners.set(i, top);
                    break;
                }
            }
        }
        WalletOwnerRemoval(_walletOwner);
    }

    @External
    public void replaceWalletOwner(Address _walletOwner, Address _newWalletOwner) {
        onlyFromWallet();
        checkOwnerExist(_walletOwner);
        checkOwnerDoesNotExist(_newWalletOwner);
        // now we can replace the owner
        for (int i = 0; i < this.owners.size(); i++) {
            if (_walletOwner.equals(this.owners.get(i))) {
                this.owners.set(i, _newWalletOwner);
                break;
            }
        }
        WalletOwnerRemoval(_walletOwner);
        WalletOwnerAddition(_newWalletOwner);
    }

    @External
    public void changeRequirement(BigInteger _required) {
        onlyFromWallet();
        checkRequirement(this.owners.size(), _required);
        this.required.set(_required);
        RequirementChange(_required);
    }

    /**
     * Allows an owner to submit and confirm a transaction.
     */
    @External
    public void submitTransaction(Address _destination,
                                  @Optional String _method,
                                  @Optional String _params,
                                  @Optional BigInteger _value,
                                  @Optional String _description) {
        checkOwnerExist(Context.getCaller());
        Context.require(_value == null || _value.signum() >= 0);
        BigInteger transactionId = addTransaction(_destination, _method, _params, _value, _description);
        confirmTransaction(transactionId);
    }

    /**
     * Allows an owner to confirm a transaction.
     */
    @External
    public void confirmTransaction(BigInteger _transactionId) {
        Address sender = Context.getCaller();
        checkOwnerExist(sender);
        checkTransactionExist(_transactionId);
        checkNotConfirmed(_transactionId, sender);
        // set confirmation true for the sender
        this.confirmations.at(_transactionId).set(sender, true);
        Confirmation(sender, _transactionId);
        executeTransaction(_transactionId);
    }

    /**
     * Allows an owner to revoke a confirmation for a transaction.
     */
    @External
    public void revokeTransaction(BigInteger _transactionId) {
        Address sender = Context.getCaller();
        checkOwnerExist(sender);
        checkTransactionExist(_transactionId);
        checkNotExecuted(_transactionId);
        checkConfirmed(_transactionId, sender);
        // set confirmation false for the sender
        this.confirmations.at(_transactionId).set(sender, false);
        Revocation(sender, _transactionId);
    }

    /*
     * Read-only methods
     */
    @External(readonly=true)
    public BigInteger getRequirement() {
        return this.required.get();
    }

    @External(readonly=true)
    public List<Address> getWalletOwners() {
        int len = this.owners.size();
        Address[] array = new Address[len];
        for (int i = 0; i < len; i++) {
            array[i] = this.owners.get(i);
        }
        return List.of(array);
    }

    @External(readonly=true)
    public int getConfirmationCount(BigInteger _transactionId) {
        int count = 0;
        for (int i = 0; i < this.owners.size(); i++) {
            if (this.confirmations.at(_transactionId).getOrDefault(this.owners.get(i), false)) {
                count++;
            }
        }
        return count;
    }

    @External(readonly=true)
    public List<Address> getConfirmations(BigInteger _transactionId) {
        int len = this.owners.size();
        Address[] array = new Address[len];
        int count = 0;
        for (int i = 0; i < len; i++) {
            Address owner = this.owners.get(i);
            if (this.confirmations.at(_transactionId).getOrDefault(owner, false)) {
                array[count++] = owner;
            }
        }
        Address[] confirmations = new Address[count];
        System.arraycopy(array, 0, confirmations, 0, count);
        return List.of(confirmations);
    }

    @External(readonly=true)
    public BigInteger getTransactionCount(boolean _pending, boolean _executed) {
        BigInteger count = BigInteger.ZERO;
        BigInteger total = this.transactionCount.getOrDefault(BigInteger.ZERO);
        while (total.signum() > 0) {
            total = total.subtract(BigInteger.ONE);
            Transaction transaction = this.transactions.get(total);
            if (_pending && !transaction.executed() || _executed && transaction.executed()) {
                count = count.add(BigInteger.ONE);
            }
        }
        return count;
    }

    @External(readonly=true)
    public Map<String, String> getTransactionInfo(BigInteger _transactionId) {
        Transaction transaction = this.transactions.get(_transactionId);
        if (transaction == null) {
            return Map.of();
        }
        return transaction.toMap(_transactionId);
    }

    @External(readonly=true)
    public List<BigInteger> getTransactionIds(BigInteger _offset, BigInteger _count,
                                              boolean _pending, boolean _executed) {
        Context.require(_offset.signum() >= 0 && _count.signum() >= 0);
        BigInteger total = this.transactionCount.getOrDefault(BigInteger.ZERO);
        if (_offset.add(_count).compareTo(total) > 0) {
            _count = total.subtract(_offset);
        }
        if (_count.signum() <= 0) {
            return List.of();
        }
        BigInteger[] entries = new BigInteger[_count.intValue()];
        int index = 0;
        for (int i = 0; _count.signum() > 0; i++) {
            _count = _count.subtract(BigInteger.ONE);
            BigInteger transactionId = _offset.add(BigInteger.valueOf(i));
            Transaction transaction = this.transactions.get(transactionId);
            if (_pending && !transaction.executed() || _executed && transaction.executed()) {
                entries[index++] = transactionId;
            }
        }
        if (index < entries.length) {
            BigInteger[] tmp = new BigInteger[index];
            System.arraycopy(entries, 0, tmp, 0, index);
            entries = tmp;
        }
        return List.of(entries);
    }

    @External(readonly=true)
    public List<Map<String, String>> getTransactionList(BigInteger _offset, BigInteger _count,
                                                        boolean _pending, boolean _executed) {
        Context.require(_offset.signum() >= 0 && _count.signum() >= 0);
        BigInteger total = this.transactionCount.getOrDefault(BigInteger.ZERO);
        if (_offset.add(_count).compareTo(total) > 0) {
            _count = total.subtract(_offset);
        }
        if (_count.signum() <= 0) {
            return List.of();
        }
        @SuppressWarnings("unchecked")
        Map<String, String>[] entries = new Map[_count.intValue()];
        int index = 0;
        for (int i = 0; _count.signum() > 0; i++) {
            _count = _count.subtract(BigInteger.ONE);
            BigInteger transactionId = _offset.add(BigInteger.valueOf(i));
            Transaction transaction = this.transactions.get(transactionId);
            if (_pending && !transaction.executed() || _executed && transaction.executed()) {
                entries[index++] = transaction.toMap(transactionId);
            }
        }
        if (index < entries.length) {
            @SuppressWarnings("unchecked")
            Map<String, String>[] tmp = new Map[index];
            System.arraycopy(entries, 0, tmp, 0, index);
            entries = tmp;
        }
        return List.of(entries);
    }

    /*
     * Assertion methods
     */
    private void onlyFromWallet() {
        assert(Context.getAddress() != null);
        Context.require(Context.getAddress().equals(Context.getCaller()));
    }

    private void checkRequirement(int ownerCount, BigInteger required) {
        int _required = required.intValue();
        Context.require((ownerCount <= MAX_OWNER_COUNT)
                && (_required <= ownerCount)
                && (_required > 0));
    }

    private void checkOwnerExist(Address owner) {
        //TODO: iteration is not efficient. Consider to use a Map.
        for (int i = 0; i < this.owners.size(); i++) {
            if (owner.equals(this.owners.get(i))) {
                return;
            }
        }
        Context.revert(100, "Owner not exist");
    }

    private void checkOwnerDoesNotExist(Address owner) {
        //TODO: iteration is not efficient. Consider to use a Map.
        for (int i = 0; i < this.owners.size(); i++) {
            if (owner.equals(this.owners.get(i))) {
                Context.revert(101, "Owner already exists");
            }
        }
    }

    private void checkTransactionExist(BigInteger transactionId) {
        Context.require(this.transactions.get(transactionId) != null);
    }

    private void checkConfirmed(BigInteger transactionId, Address sender) {
        Context.require(this.confirmations.at(transactionId).getOrDefault(sender, false));
    }

    private void checkNotConfirmed(BigInteger transactionId, Address sender) {
        Context.require(!this.confirmations.at(transactionId).getOrDefault(sender, false));
    }

    private void checkNotExecuted(BigInteger transactionId) {
        Transaction transaction = this.transactions.get(transactionId);
        Context.require(!transaction.executed());
    }

    /*
     * Internal methods
     */
    private BigInteger addTransaction(Address destination, String method, String params,
                                      BigInteger value, String description) {
        BigInteger transactionId = this.transactionCount.getOrDefault(BigInteger.ZERO);
        this.transactions.set(transactionId,
                new Transaction(destination, method, params, value, description));
        this.transactionCount.set(transactionId.add(BigInteger.ONE));
        Submission(transactionId);
        return transactionId;
    }

    private void executeTransaction(BigInteger transactionId) {
        if (isConfirmed(transactionId)) {
            Transaction transaction = this.transactions.get(transactionId);
            Context.require(!transaction.executed());
            if (externalCall(transaction)) {
                transaction.setExecuted(true);
                // we need to set the transaction again since we changed the executed status
                this.transactions.set(transactionId, transaction);
                Execution(transactionId);
            } else {
                ExecutionFailure(transactionId);
            }
        }
    }

    private boolean isConfirmed(BigInteger transactionId) {
        int count = 0;
        int required = this.required.get().intValue();
        for (int i = 0; i < this.owners.size(); i++) {
            if (this.confirmations.at(transactionId).getOrDefault(this.owners.get(i), false)) {
                count++;
            }
        }
        // execute the transaction only if the confirmed count exactly matches the required
        return count == required;
    }

    private boolean externalCall(Transaction transaction) {
        try {
            Context.call(transaction.value(), transaction.destination(),
                    transaction.method(), transaction.getConvertedParams());
            return true;
        } catch (Exception e) {
            Context.println("[Exception] " + e.getMessage());
            return false;
        }
    }

    /*
     * Events
     */
    @EventLog(indexed=1)
    protected void WalletOwnerAddition(Address _walletOwner) {}

    @EventLog(indexed=1)
    protected void WalletOwnerRemoval(Address _walletOwner) {}

    @EventLog
    protected void RequirementChange(BigInteger _required) {}

    @EventLog(indexed=1)
    protected void Submission(BigInteger _transactionId) {}

    @EventLog(indexed=2)
    protected void Confirmation(Address _sender, BigInteger _transactionId) {}

    @EventLog(indexed=2)
    protected void Revocation(Address _sender, BigInteger _transactionId) {}

    @EventLog(indexed=1)
    protected void Execution(BigInteger _transactionId) {}

    @EventLog(indexed=1)
    protected void ExecutionFailure(BigInteger _transactionId) {}

    @EventLog(indexed=1)
    protected void Deposit(Address _sender, BigInteger _value) {}

    @EventLog(indexed=1)
    protected void DepositToken(Address _sender, BigInteger _value, byte[] _data) {}
}
